src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+        //
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.on_line(line);
+            }
+        }
+        self.after_event();
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(mut opt: Options) -> Self {
+        Self {
+            in_event: false,
+            skip_stack: false,
+            stack: VecDeque::default(),
+            count: 0,
+            occurrences: HashMap::default(),
+            pname: None,
+            opt,
+        }
+    }
+}
+
+impl Folder {
+    fn on_line(&mut self, line: &str) {
+        self.in_event = true;
+        self.on_stack_line(line)
+    }
+
+    fn stack_line_parts(line: &str) -> Option<(&str, &str)> {
+        let mut line = line.splitn(2, '+');
+        let path = line.next()?;
+
+        let addr = if let Some(addr) = line.next() {
+            addr
+        } else {
+            ""
+        };
+        Some((path, addr))
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        if let Ok(count) = line.parse::<usize>() {
+            self.count = count;
+        } else {
+            if let Some((func, addr)) = Folder::stack_line_parts(line) {
+                self.stack.push_front(func.to_string());
+            } else {
+                warn!("weird stack line: {}", line);
+            }
+        }
+    }
+
+    fn after_event(&mut self) {
+        if !self.skip_stack && self.in_event {
+            // allocate a string that is long enough to hold the entire stack string
+            let mut len = self.stack.iter().fold(0, |a, s| a + s.len() + 1);
+            if let Some(pname) = &self.pname {
+                len += pname.len();
+            }
+            let mut stack_str = String::with_capacity(len);
+
+            // add the comm name
+            let mut first = true;
+            if let Some(pname) = &self.pname {
+                stack_str.push_str(pname);
+                first = false;
+            }
+
+            // add the other stack entries (if any)
+            for e in self.stack.drain(..) {
+                if first {
+                    first = false
+                } else {
+                    stack_str.push_str(";");
+                }
+                stack_str.push_str(&e);
+            }
```

Comment:
Do we need to also "[trim leaf offsets](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L73L74)" here?

Reply:
this is the address removeal

Reply:
I'm not entirely sure what you mean? Upstream removes the address in _two_ places: [here](https://github.com/brendangregg/FlameGraph/blob/1b1c6deede9c33c5134c920bdb7a44cc5528e9a7/stackcollapse.pl#L73L74) and [here](https://github.com/brendangregg/FlameGraph/blob/1b1c6deede9c33c5134c920bdb7a44cc5528e9a7/stackcollapse.pl#L85). How come we only have to do it in one place?

Reply:
The upstream script seems to be bugged there, it removes the offset `if $includeoffset` and `unless $includeoffset`.

Reply:
I missread that, it jsut removes the last one if `includeoffset` is set to true.

Reply:
re-added